local ReplicatedStorage = game:GetService("ReplicatedStorage")

local types = require(ReplicatedStorage:WaitForChild("global"):WaitForChild("types"))
local objects = require(ReplicatedStorage:WaitForChild("global"):WaitForChild("utils"):WaitForChild("objects"))

export type LifeStats = types.LifeStats
export type SlotSummary = types.SlotSummary
export type SlotData = types.SlotData
export type Account = types.Account
export type TransactionFunction = (player: Player, () -> ()) -> ()
export type Transaction = (txFor: TransactionFunction) -> ()
export type ProfileData = {
	account: Account,
	migrationVersion: number,
}

local DEFAULT_SLOT_COUNT = 3

local function makeEmptySlotSummary(slotId: number): SlotSummary
	return {
		SlotId = slotId,
		Exists = false,
		Locked = false,
		Name = "Empty Slot",
		Age = 0,
		CurrentMonth = 1,
		MonthsIntoAge = 0,
		Grade = "None",
		Generation = 1,
		FirstName = "Unknown",
		FamilyName = "Unknown",
		Origin = "Unknown",
		Clan = "Unknown",
		Gender = "Unknown",
		CursedTechnique = "None",
		LastPlayed = 0,
		LastSaved = 0,
	}
end

local defaultSettings: types.Settings = {
	MusicVolume = 1,
	SFXVolume = 1,
	MuteMusic = false,
	MuteSFX = false,
	FastText = false,
	HighContrastText = false,
	ReducedMotion = false,
	ScreenShake = true,
	SkipCombatSkillChecks = false,
}

local defaultLifeStats: LifeStats = {
	DataVersion = 2,
	Name = "Unknown",
	FirstName = "Unknown",
	FamilyName = "Unknown",
	FamilyType = "Unknown",
	Gender = "Unknown",
	Age = 0,
	CurrentMonth = 1,
	MonthsIntoAge = 0,
	Health = 100,
	Happiness = 100,
	Smarts = 50,
	Alive = true,
	CursedEnergy = 0,
	Grade = "Non-Sorcerer",
	Clan = "Minor Sorcerer Family",
	Origin = "Minor Sorcerer Family",
	OriginCategory = "Family",
	CursedTechnique = "None",
	TechniqueMastery = 0,
	InnateTechnique = "None",
	TechniqueType = "None",
	TechniqueKnown = false,
	TechniqueUsable = false,
	TechniqueAwakened = false,
	InnateDormant = false,
	PlayerClanInnateTechnique = "None",
	TechniqueEvolutionStage = "Dormant",
	TechniqueEvolutionProgress = 0,
	TechniqueBranch = "None",
	TechniqueUsageCount = 0,
	TechniqueMissionUses = 0,
	UnlockedExtensions = {},
	IsCurse = false,
	CurseType = "None",
	CurseGrade = "None",
	CurseClassified = false,
	CurseThreat = 0,
	CurseThreatClass = "None",
	CurseEvolution = "None",
	CurseEvolutionProgress = 0,
	CurseMutations = {},
	CurseMutationCount = 0,
	CurseLatestMutation = "None",
	CurseHunger = 0,
	CurseTerritory = "None",
	CurseRumor = "None",
	SocietyAttention = 0,
	CurseFearLevel = 0,
	CurseTerritoryControl = 0,
	CurseRumorSpread = 0,
	CiviliansTerrified = 0,
	HuntersEscaped = 0,
	CursedResentment = 0,
	CurseRebirthEligible = false,
	CurseRebirthCount = 0,
	SorcererAttackCooldown = 0,
	SorcerersKilled = 0,
	ExorcismRisk = 0,
	TimesExorcised = 0,
	SocietyStatus = "Civilian",
	RegisteredSorcerer = false,
	Recommendations = 0,
	RecommendationProgress = 0,
	PromotionDifficulty = 1,
	WantedLevel = 0,
	CursedFingers = 0,
	CursedObjectAnchors = 0,
	CursedObjectAnchorProgress = 0,
	ReincarnationAnchor = "None",
	SocietyRecognition = 0,
	SocietyInfluence = 0,
	SocietyRank = "None",
	CouncilSeats = 0,
	CouncilCampaigns = 0,
	SocietyPolicyWins = 0,
	LastSocietyOutcome = "None",
	HeianEraEncounters = 0,
	EdoEraEncounters = 0,
	ModernEraEncounters = 0,
	EraLoreDiscoveries = 0,
	SpecialGradeThreatsSurvived = 0,
	LastEraLoreOutcome = "None",
	Fame = 0,
	Fear = 0,
	WorldAttention = 0,
	WorldPressure = 0,
	WorldReputationTier = "Unknown",
	PublicSupport = 0,
	HunterPressure = 0,
	CopycatIncidents = 0,
	WorldReputationEvents = 0,
	ReputationCleanups = 0,
	LastWorldReputationEvent = "None",
	LifePath = "Undecided",
	PathReputation = 0,
	RetiredFromSorcery = false,
	RetirementAge = -1,
	RetirementYears = 0,
	CivilianStability = 0,
	RetirementRoute = "None",
	Affiliation = "Unaffiliated",
	AffiliationRank = "None",
	AffiliationReputation = 0,
	AffiliationDuties = 0,
	JujutsuHighPopularity = 0,
	GoodwillWins = 0,
	TournamentRuns = 0,
	TournamentBestStreak = 0,
	TournamentsWon = 0,
	InJujutsuHigh = false,
	AssistantTrust = 0,
	AssistantSkill = 0,
	Money = 0,
	Reputation = 0,
	TeacherRecommendations = 0,
	AFKStabilizationUnlocked = false,
	AFKStabilizationEarned = 0,
	AFKStabilizationSessions = 0,
	MissionsCompleted = 0,
	MissionsFailed = 0,
	ActivitiesDoneThisAge = 0,
	RelationshipActionsThisAge = 0,
	RecentLifeEvents = {},
	RecentQuietTexts = {},
	Assets = {},
	Relationships = {},
	RelationshipHistory = {},
	SecretAchievements = {},
	UnlockedPremiumClans = {},
	UnlockedPremiumTechniques = {},
	WorldHistory = {},
	NotableSorcerers = {},
	UnlockedTitles = {},
	MiracleCharges = 0,
	CursedTools = {},
	ClanWarState = {},
	ClanWarHistory = {},
	ClanWarWins = 0,
	ClanWarLosses = 0,
	ClanWarActions = 0,
	ClanWarCooldown = 0,
	DojoReputation = 0,
	DojoRank = "No Dojo",
	DojoIncome = 0,
	DojoLegacy = 0,
	DojoLegacyTier = "None",
	DojoSuccessors = 0,
	DojoNamedSuccessors = 0,
	DojoLegacyCeremonies = 0,
	ChildrenInductedIntoDojo = 0,
	DojoProtectionBonus = 0,
	LastDojoLegacyAge = -1,
	StudentMissionCooldown = 0,
	DojoLevel = 1,
	DojoCondition = 100,
	DojoPrestige = 0,
	DojoFocus = "Balanced",
	DojoSparringEvents = 0,
	DojoFounderName = "Unknown",
	DojoNameConfirmed = false,
	DojoPendingLastNameChoice = false,
	DojoRecruitPool = {},
	DojoStarterRecruitLimit = 3,
	DojoStarterRecruitsUsed = 0,
	DojoRaidRisk = 0,
	DojoRaidsDefended = 0,
	DojoRaidsFailed = 0,
	DojoRaidDamageTaken = 0,
	DojoRaidStudentsInjured = 0,
	DojoRaidSuppliesLost = 0,
	DojoRecoveryDebt = 0,
	DojoRecoveryProjects = 0,
	LastDojoRaidType = "None",
	LastDojoRaidOutcome = "None",
	RogueStudentIncidents = 0,
	WardsRescued = 0,
	WardsAdopted = 0,
	RecruitRerollTokens = 0,
	ClanFounded = false,
	ClanMembers = 0,
	ClanReputation = 0,
	ClanStudentsRecruited = 0,
	ClanTrainingSessions = 0,
	FamilyReputation = 0,
	ClanLeadership = "None",
	ClanPoliticalPower = 0,
	ClanMergeRequestsSent = 0,
	ClanMergeAlliances = 0,
	ClanMergePacts = 0,
	ClanMergeRejections = 0,
	ClanRivalries = 0,
	ClanMergeAccepted = false,
	LastClanMergeOutcome = "None",
	BloodlineStrength = 0,
	DeathPaintingLineage = false,
	LegacyHistory = {},
	Generation = 1,
	Achievements = { Claimed = {} },
	Codes = { Redeemed = {} },
	Purchases = { StarterSorcererPackClaimed = false },
	Settings = defaultSettings,
}

local accountTemplate: Account = {
	selectedSlot = 1,
	slotCount = DEFAULT_SLOT_COUNT,
	extraSlotsUnlocked = false,
	giftedPasses = {},
	accountRewards = {},
	dailyLogin = {},
	slots = {
		{ exists = false, summary = makeEmptySlotSummary(1), stats = nil },
		{ exists = false, summary = makeEmptySlotSummary(2), stats = nil },
		{ exists = false, summary = makeEmptySlotSummary(3), stats = nil },
	},
}

return {
	template = {
		account = accountTemplate,
		migrationVersion = 0,
	} :: ProfileData,
	accountSchema = types.accountSchema,
	lifeStatsSchema = types.lifeStatsSchema,
	slotSummarySchema = types.slotSummarySchema,
	defaultLifeStats = defaultLifeStats,
	defaultSettings = defaultSettings,
	makeEmptySlotSummary = makeEmptySlotSummary,
	DEFAULT_SLOT_COUNT = DEFAULT_SLOT_COUNT,
}
